README.md
# SVG-to-PNG Converter Using Chromium

You have a SVG file. For whatever reason, you need a PNG. **svg2png** can help.

```js
const fs = require("pn/fs"); // https://www.npmjs.com/package/pn
const svg2png = require("svg2png");

svg2png("~/test.svg")
    .then(buffer => fs.writeFile("dest.png", buffer))
    .catch(e => console.error(e));
```

In the above example, we use the `width` and `height` attributes specified in the SVG file to automatically determine the size of the SVG. You can also explicitly set the size:

```js
svg2png(url, { width: 300, height: 400 })
    .then(buffer => ...)
    .catch(e => console.error(e));
```

This is especially useful for images without `width` or `height`s. You can even specify just one of them and (if the image has an appropriate `viewBox`) the other will be set to scale.


<<<<<<< HEAD
## Sync variant

There's also a sync variant, for use in your shell scripts:

```js
const outputBuffer = svg2png.sync(url, options);
```

=======
>>>>>>> 61541144915ed9c54bb680e53fa42faba9e22abc
## How the conversion is done

svg2png uses [puppeteer](https://github.com/GoogleChrome/puppeteer) to render your SVGs using a headless Chromium instance. I have found this to produce much more accurate renderings than other solutions like GraphicsMagick or Inkscape. Plus, it's easy to install cross-platform since it downloads Chromium on install so you don't even need to have it in your `PATH`.

## Exact resizing behavior

Previous versions of svg2png attempted to infer a good size based on the `width`, `height`, and `viewBox` attributes. As of our 3.0 release, we attempt to stick as close to the behavior of loading a SVG file in your browser as possible. The rules are:

- Any `width` or `height` attributes that are in percentages are ignored and do not count for the subsequent rules.
- The dimensions option `{ width, height }` overrides any `width` or `height` attributes in the SVG file, including for the subsequent rules. If a key is missing from the dimensions object (i.e. `{ width }` or `{ height }`) the corresponding attribute in the SVG file will be deleted.
- `width` and `height` attributes without a `viewBox` attribute cause the output to be of those dimensions; this might crop the image or expand it with empty space to the bottom and to the right.
- `width` and/or `height` attributes with a `viewBox` attribute cause the image to scale to those dimensions. If the ratio does not match the `viewBox`'s aspect ratio, the image will be expanded and centered with empty space in the extra dimensions. When a `viewBox` is present, one of either `width` or `height` can be omitted, with the missing one inferred from the `viewBox`'s aspect ratio.
- When there are neither `width` nor `height` attributes, the promise rejects.

One thing to note is that svg2png does not and cannot stretch your images to new aspect ratios.

## CLI

This package comes with a CLI version as well; you can install it globally with `npm install svg2png -g`. Use it as follows:

```
$ svg2png --help
Converts SVGs to PNGs, using Chromium

svg2png input.svg [--output=output.png] [--width=300] [--height=150]

Options:
  -o, --output  The output filename; if not provided, will be inferred  [string]
  -w, --width   The output file width, in pixels                        [string]
  -h, --height  The output file height, in pixels                       [string]
  --help        Show help                                              [boolean]
  --version     Show version number                                    [boolean]
```

## Node.js requirements

svg2png uses the latest in ES2017 features, and as such requires a recent version of Node.js. Only the 8.x series onward is supported.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project
adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]


## [0.0.1] - Date of release
Description


[Unreleased]: https://github.com/iOffice/svg2png/compare/0.0.1...HEAD
[0.0.1]: https://github.com/iOffice/svg2png/compare/<hash/tag>...0.0.1


src/test/util.ts
import * as path from 'path';
import * as fs from 'fs';
import { Stream } from 'stream';
import fileURL = require('file-url');
import { ReadStream } from 'fs';
import { PNG } from 'pngjs';

const rel = (x: string) => path.resolve(__dirname, x);
const streamifier = require('streamifier');

interface ITest {
  name: string;
  file: string;
  options?: any;
  'notes to validate': string;
  includeURL?: string;
  includeFilename?: string;
}

type StrBuffStream = string | Buffer | Stream;

function normalizeTests(tests: ITest[]) {
  return tests.map(test => {
    const normalized = { ...test };
    const filename = rel(`../../src/test/inputs/${test.file}`);

    if (normalized.options || normalized.includeFilename || normalized.includeURL) {
      normalized.options = { ...test.options };

      if (normalized.includeFilename) {
        normalized.options.filename = filename;
        delete normalized.includeFilename;
      }
      if (normalized.includeURL) {
        normalized.options.url = fileURL(filename);
        delete normalized.includeURL;
      }
    }

    normalized.file = filename;
    return normalized;
  });
}


function toStream(obj: StrBuffStream): Promise<ReadStream> {
  return new Promise((resolve, reject) => {
    let result;
    if (typeof obj === 'string') {
      result = fs.createReadStream(obj).once('error', reject);
    }
    if (obj instanceof Buffer) {
      result = streamifier.createReadStream(obj).once('error', reject);
    }
    if (obj instanceof Stream) {
      result = obj;
    }
    resolve(result);
  });
}

function parsePNG(obj: StrBuffStream): Promise<PNG> {
  return new Promise(async (resolve, reject) => {
    try {
      const stream = await toStream(obj);
      stream.pipe(new PNG())
        .once('error', reject)
        .on('parsed', function onParsed() {
          resolve(this);
        });
    } catch (err) {
      reject(err);
    }
  });
}

async function pngCompare(obj1: StrBuffStream, obj2: StrBuffStream) {
  const png1 = await parsePNG(obj1);
  const png2 = await parsePNG(obj2);
  const dim1 = [png1.width, png1.height];
  const dim2 = [png2.width, png2.height];
  if (png1.data.length !== png2.data.length) {
    return Promise.reject(new Error(`Dimension mismatch: [${dim1}] vs [${dim2}].`));
  }

  const writeStream = new PNG({ width: dim1[0], height: dim1[1] });
  const data = writeStream.data;
  let i = 0;
  let count = 0;
  while (png1.data[i] != null) {
    if (
      png1.data[i] !== png2.data[i] ||
      png1.data[i + 1] !== png2.data[i + 1] ||
      png1.data[i + 2] !== png2.data[i + 2] ||
      png1.data[i + 3] !== png2.data[i + 3]
    ) {
      count += 1;
      data[i] = png2.data[i];
      data[i + 1] = png2.data[i + 1];
      data[i + 2] = png2.data[i + 2];
      data[i + 3] = png2.data[i + 3];
    }
    i += 4;
  }

  return {
    diffPixels: count,
    totalPixels: dim1[0]*dim1[1],
    png: writeStream.pack(),
  };
}


export {
  ITest,
  normalizeTests,
  toStream,
  pngCompare,
};


src/test/index.ts
import * as fs from 'fs';
import * as path from 'path';
import { expect } from 'chai';
import { ITest, normalizeTests, pngCompare } from './util';
import { Svg2png, svg2png } from '../main/svg2png';

const rel = (x: string) => path.resolve(__dirname, x);
const successTests = normalizeTests(require('../../src/test/success-tests.json'));
const failureTests = normalizeTests(require('../../src/test/failure-tests.json'));


describe('async', () => {
  describe('should fulfill', () => successTests.forEach(successTest));
  describe("should reject", () => failureTests.forEach(failureTest));

  after(async () => {
    await Svg2png.closePool();
  });
});


function successTest(test: ITest, index: number) {
  it(test.name, async () => {
    const expected = fs.readFileSync(rel(`../../src/test/success-tests/${index}.png`));
    const actual = await svg2png({
      url: test.file,
      ...test.options,
    });
    const result = await pngCompare(expected, actual);
    const ratio = result.diffPixels/result.totalPixels;
    try {
      expect(ratio).to.be.lessThan(0.01, 'More than 1% of the pixels are different');
    } catch (err) {
      result.png.pipe(fs.createWriteStream(`err_${index}.png`));
      throw err;
    }
  });
}

function failureTest(test: ITest) {
  it(test.name, async () => {
    try {
      await svg2png({
        url: test.file,
        ...test.options,
      });
    } catch (err) {
      return expect(err.message).to.include(test['expectedErrorSubstring']);
    }
    throw new Error('Did not get the exception');
  });
}



lib/svg2png.js
/* eslint-env browser */

"use strict";
const fileURL = require("file-url");
const fs = require("pn/fs");
const puppeteer = require("puppeteer");
const createPuppeteerPool = require("puppeteer-pool");
const tmp = require("tmp");
const sharp = require("sharp");

const pool = createPuppeteerPool({
    min: 1,
    max: 2,
    maxUses: 5,
    idleTimeoutMillis: 5000,
    puppeteerArgs: { args: ['--no-sandbox', '--disable-setuid-sandbox']}
});
let pages = {};
let idCounter = 0;

module.exports = async (source, options) => {
    try {
        options = parseOptions(options);

        const id = ++idCounter;
        const timeout = new Promise((resolve, reject) => {
            setTimeout(() => {
                if (pages[id]) {
                    pages[id].close();
                    reject(new Error("Timeout rasterizing SVG"));
                } else {
                    resolve()
                }
            }, 300000);
        });
        const convertPromise = convertBuffer(id, source, options);

        return await Promise.race([convertPromise, timeout]);
    } catch (e) {
        throw e;
    }
};

async function convertBuffer(id, source, options) {
    const output = await pool.use(async (browser) => {
        try {
            const page = await browser.newPage();
            pages[id] = page;

            await page.goto(source, {
                waitUntil: ['load', 'domcontentloaded', 'networkidle0'],
                timeout: options.timeout
            });

            await setDimensions(page, options);

            const dimensions = await getDimensions(page);
            if (!dimensions) {
                throw new Error("Width or height could not be determined from either the source file or the supplied " +
                                "dimensions");
            }

            const width = Math.round(dimensions.width);
            const fullHeight = Math.round(dimensions.height);
            await page.setViewport({ width, height: fullHeight });

            /* Chrome can only render 16 1024 by 1024 blocks, or 256 256x256
             * in most cases. Solution is to caclulate the number of blocks
             * that can fit in a screenshot, and stich them together.
             * See crbug.com/770769 for more info
             */
            const blocksPerRow = Math.floor(width / 256) - 1; // Subtract 1 row for safety
            const blocksPerCol = Math.floor(256 / blocksPerRow);
            const maxScreenshotHeight = blocksPerCol * 256; 

            if (fullHeight <= maxScreenshotHeight) { // no point in doing all this work if it fits
                return await page.screenshot({
                    fullPage: true,
                    omitBackground: true
                });
            } else {
                /* Since our screenshots span the full width, if we want to stack them
                 * together, we can just put append the raw data, throw it into
                 * a buffer, and tell sharp the correct width, height, and channel
                 */
                const chunks = []; 
                for (let ypos = 0; ypos < fullHeight; ypos += maxScreenshotHeight) {
                    const height = Math.min(fullHeight - ypos, maxScreenshotHeight);
                    const screenshot = await page.screenshot({
                        clip: {
                            x: 0,
                            y: ypos,
                            width,
                            height
                        },
                        omitBackground: true
                    });
                   const buffer = await sharp(screenshot).raw().toBuffer();
                   chunks.push(buffer);
                }
                const channels = 4;
                const bufferSize = width * maxScreenshotHeight * channels;
                const composite = Buffer.allocUnsafe(bufferSize * chunks.length);
                chunks.forEach((s, i) => s.copy(composite, i * bufferSize))
                return await sharp(composite, {
                    raw: {
                        width,
                        height: fullHeight,
                        channels
                    }
                }).png().toBuffer();
            }
        } finally {
            const page = pages[id];
            if (page) {
                try {
                    delete pages[id];
                    await page.close()
                } catch (err) {
                    console.log("Page close error:", err);
                }
            }
        }
    })

    return output;
}

async function getDimensions(page) {
    return await page.evaluate(() => {
        const el = document.querySelector("svg");

        const widthIsPercent = (el.getAttribute("width") || "").endsWith("%");
        const heightIsPercent = (el.getAttribute("height") || "").endsWith("%");
        const width = !widthIsPercent && parseFloat(el.getAttribute("width"));
        const height = !heightIsPercent && parseFloat(el.getAttribute("height"));

        if (width && height) {
            return { width, height };
        }

        const viewBoxWidth = el.viewBox.animVal.width;
        const viewBoxHeight = el.viewBox.animVal.height;

        if (width && viewBoxHeight) {
            return { width, height: width * viewBoxHeight / viewBoxWidth };
        }

        if (height && viewBoxWidth) {
            return { width: height * viewBoxWidth / viewBoxHeight, height };
        }

        return null;
    });
}

async function setDimensions(page, dimensions) {
    if (dimensions.width === undefined && dimensions.height === undefined && dimensions.scale === undefined) {
        return;
    }

    await page.evaluate(({ width, height, scale }) => {
        const el = document.querySelector("svg");
        if (!el) {
            return;
        }

        if (width !== undefined) {
            el.setAttribute("width", `${width}px`);
        } else {
            el.removeAttribute("width");
        }

        if (height !== undefined) {
            el.setAttribute("height", `${height}px`);
        } else {
            el.removeAttribute("height");
        }

        if (scale !== undefined) {
            var viewBoxWidth = el.viewBox.animVal.width;
            var viewBoxHeight = el.viewBox.animVal.height;
            var scaledWidth = viewBoxWidth / scale;
            var scaledHeight = viewBoxHeight / scale;
            el.setAttribute("width", scaledWidth + "px");
            el.setAttribute("height", scaledHeight + "px");
            el.removeAttribute("clip-path");
            /* It might eventually be necessary to scale the clip path of the root svg element
            var clipPathEl = document.querySelector("svg > clipPath > path");
            clipPathEl.setAttribute("d", `M0 0v${scaledHeight}h${scaledWidth}V0z`)
            */
        }
    }, dimensions);
}

function parseOptions(options) {
    options = Object.assign({}, options);

    if (typeof options.width === "string") {
        options.width = parseInt(options.width);
    }
    if (typeof options.height === "string") {
        options.height = parseInt(options.height);
    }
    if (typeof options.scale === "string") {
        options.scale = parseFloat(options.scale);
    }

    if (options.filename !== undefined && options.url !== undefined) {
        throw new Error("Cannot specify both filename and url options");
    }

    // Convert filename option to url option
    if (options.filename !== undefined) {
        options.url = fileURL(options.filename);
        delete options.filename;
    }

    return options;
}


lib/converter.js
/* eslint-env phantomjs */
/* eslint-disable no-console, no-var, prefer-arrow-callback, object-shorthand */
"use strict";

var webpage = require("webpage");
var system = require("system");

var HTML_PREFIX = "<!DOCTYPE html><style>html, body { margin: 0; padding: 0; } " +
                  "svg { position: absolute; top: 0; left: 0; }</style>";

if (system.args.length !== 3) {
    console.error("Usage: converter.js url options");
    phantom.exit();
} else {
    convert(system.args[1], system.args[2]);
}

function convert(url, options) {
    try {
        options = JSON.parse(options);
    } catch (e) {
        console.error("Unable to parse options.");
        console.error(e);
        phantom.exit();
        return;
    }

    var page = webpage.create();

    page.onResourceError = function(error) {
        page.reason = error.errorString;
    };

    page.onLoadFinished = function (status) {
        if (status !== "success") {
            console.error("Unable to load the source file: " + page.reason);
            phantom.exit();
            return;
        }

        try {
            if (options.width !== undefined || options.height !== undefined || options.scale !== undefined) {
                setSVGDimensions(page, options.width, options.height, options.scale);
            }

            var dimensions = getSVGDimensions(page);
            if (!dimensions) {
                console.error("Width or height could not be determined from either the source file or the supplied " +
                              "dimensions");
                phantom.exit();
                return;
            }

            setSVGDimensions(page, dimensions.width, dimensions.height);

            page.viewportSize = {
                width: dimensions.width,
                height: dimensions.height
            };
            page.clipRect = {
                top: 0,
                left: 0,
                width: dimensions.width,
                height: dimensions.height
            };
        } catch (e) {
            console.error("Unable to calculate or set dimensions.");
            console.error(e);
            phantom.exit();
            return;
        }

        page.evaluate(function() {
            var previousT = 0;
            var t = 0;
            var fpsList = [];

            function stdev(values) {
                var mean = values.reduce(function(a, b) { return a + b; }) / values.length;
                var devSquared = values.map(function(a) { return Math.pow(a - mean, 2); }).reduce(function(a, b) { return a + b; });
                return Math.sqrt(devSquared / (values.length - 1));
            }

            function calcFPS(now) {
                t = now;
                var fps = 1000 / (t - previousT);
                fpsList.push(fps);
                var std = stdev(fpsList.slice(fpsList.length - 50, fpsList.length));
                if (std <= 5) {
                    console.log("done");
                    return;
                }
                previousT = t;
                requestAnimationFrame(calcFPS);
            }

            requestAnimationFrame(calcFPS);
        });

        function saveResult() {
            var result = "data:image/png;base64," + page.renderBase64("PNG");
            system.stdout.write(result);
            phantom.exit();
        }

        var timeout = setTimeout(function() {
            saveResult();
        }, 20000);

        page.onConsoleMessage = function(msg) {
            if (msg.indexOf("done") >= 0) {
                clearTimeout(timeout);
                saveResult();
            }
        }
    };

    // PhantomJS will always render things empty if you choose about:blank, so that's why the different default URL.
    // PhantomJS's setContent always assumes HTML, not SVG, so we have to massage the page into usable HTML first.
    page.open(url);
}

function setSVGDimensions(page, width, height, scale) {
    if (width === undefined && height === undefined && scale === undefined) {
        return;
    }

    page.evaluate(function (widthInside, heightInside, scaleInside) {
        /* global document: false */
        var el = document.querySelector("svg");

        if (widthInside !== undefined) {
            el.setAttribute("width", widthInside + "px");
        } else {
            el.removeAttribute("width");
        }

        if (heightInside !== undefined) {
            el.setAttribute("height", heightInside + "px");
        } else {
            el.removeAttribute("height");
        }

        if (scaleInside !== undefined) {
            var viewBoxWidth = el.viewBox.animVal.width;
            el.setAttribute("width", Math.round(viewBoxWidth / scaleInside) + "px");
        }
    }, width, height, scale);
}

function getSVGDimensions(page) {
    return page.evaluate(function () {
        /* global document: false */

        var el = document.querySelector("svg");

        var widthIsPercent = /%\s*$/.test(el.getAttribute("width") || ""); // Phantom doesn't have endsWith
        var heightIsPercent = /%\s*$/.test(el.getAttribute("height") || "");
        var width = !widthIsPercent && parseFloat(el.getAttribute("width"));
        var height = !heightIsPercent && parseFloat(el.getAttribute("height"));

        if (width && height) {
            return { width: width, height: height };
        }

        var viewBoxWidth = el.viewBox.animVal.width;
        var viewBoxHeight = el.viewBox.animVal.height;

        if (width && viewBoxHeight) {
            return { width: width, height: width * viewBoxHeight / viewBoxWidth };
        }

        if (height && viewBoxWidth) {
            return { width: height * viewBoxWidth / viewBoxHeight, height: height };
        }

        return null;
    });
}


src/main/puppeteer-pool.ts
import { Browser, LaunchOptions } from 'puppeteer';
import { Pool, Options, Factory } from 'generic-pool';
import * as puppeteer from 'puppeteer';
import * as genericPool from 'generic-pool';

interface IPuppeteerPoolConfig {
  maxUses: number;
  validator: (instance: Browser) => Promise<boolean>;
}

const USE_COUNT = Symbol('useCount');

/**
 * Originally a fork of https://github.com/latesh/puppeteer-pool.
 */
function createPuppeteerPool(
  config: IPuppeteerPoolConfig,
  puppeteerlaunchOptions: LaunchOptions,
  genericPoolConfig: Options,
): Pool<Browser> {
  const factory: Factory<Browser> = {
    async create() {
      try {
        const browser = await puppeteer.launch(puppeteerlaunchOptions);
        browser[USE_COUNT] = 0;
        return browser;
      } catch (err) {
        return Promise.reject(err);
      }
    },
    async destroy(browser: Browser) {
      try {
        await browser.close();
      } catch (err) {
      }
      return undefined;
    },
    async validate(browser: Browser) {
      try {
        const valid = await config.validator(browser);
        return valid && (config.maxUses <= 0 || browser[USE_COUNT] < config.maxUses);
      } catch (err) {
        return Promise.reject(err);
      }
    },
  };

  const pool = genericPool.createPool<Browser>(factory, genericPoolConfig);
  const genericAcquire = pool.acquire.bind(pool);

  pool.acquire = async () => {
    try {
      const browser = await genericAcquire();
      browser[USE_COUNT] += 1;
      return browser;
    } catch (err) {
      return Promise.reject(err);
    }
  };
  pool.use = async (cb: (browser: Browser) => any) => {
    let browser;
    try {
      browser = await pool.acquire();
      const result = await cb(browser);
      pool.release(browser);
      return result;
    } catch (err) {
      if (browser) {
        pool.release(browser);
      }
      return Promise.reject(err);
    }
  };

  return pool;
}

export {
  createPuppeteerPool,
};


src/main/svg2png-cli.ts
#!/usr/bin/env node
import * as fs from 'fs';
import * as path from 'path';
import * as yargs from 'yargs';
import { svg2png } from './svg2png';

async function main() {
  const name = 'svg2png';
  const argv = yargs
    .usage(`${name} input.svg ` +
      "[--output=output.png] [--width=300] [--height=150]")
    .option('o', {
      alias: 'output',
      type: 'string',
      describe: 'The output filename; if not provided, will be inferred',
    })
    .option('w', {
      alias: 'width',
      type: 'string',
      describe: 'The output file width, in pixels',
    })
    .option('h', {
      alias: 'height',
      type: 'string',
      describe: 'The output file height, in pixels',
    })
    .option('s', {
      alias: 'scale',
      type: 'string',
      describe: 'The inverted scale to use fo the output file',
    })
    .option('d', {
      alias: 'debug',
      type: 'boolean',
      describe: 'Print debug messages.',
    })
    .demand(1)
    .help(false)
    .version()
    .argv;
  try {
    const inputFilename = argv._[0];
    const outputFilename = argv.output || path.basename(inputFilename, '.svg') + ".png";
    const output = await svg2png({
      width: +argv.width,
      height: +argv.height,
      scale: +argv.scale,
      url: inputFilename,
      debug: !!argv.debug,
    });
    fs.writeFileSync(outputFilename, output);
    process.exit(0);
  } catch (e) {
    if (e.stack) {
      process.stderr.write(`${e.stack}\n`);
    } else {
      process.stderr.write(`${e}\n`);
    }
    process.exit(1);
  }
}

main();
